### ISSUE TITLE
NixOS: `textnets` isn't building in nixpkgs, causing other builds to fail (e.g. spacy)

### ISSUE THREAD

**Issue by sarahec** [jboynyc/textnets#64]
TL;dr nixpkgs has later version of `igraph` and `leidenalg` than textnets 0.9.3 is willing to accept, thus `textnets` isn't building. There are options; I'm looking for your preference.

From the 0.9.3 [build log](https://hydra.nixos.org/build/250761554/log/tail):

```console
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for textnets-0.9.3-cp311-cp311-manylinux_2_38_x86_64.whl
  - igraph<0.11.0,>=0.10.5 not satisfied by version 0.11.4
  - leidenalg<0.10.0,>=0.9.0 not satisfied by version 0.10.2
```

Options:

1. One of us can issue a patch to the 0.9.3 build that relaxes the dependency requirements. (This applies substitutions to `pyproject.toml` before building), or
2. One of us can submit an update patch to build 0.9.5. Note this also fails a dependency check as pyarrow is now at 11.0. The options here are 1) submit a tiny PR here to bump that dependency, then use that commit in the nixpkgs update, or 2) apply a substitution to pyproject to allow 11.0.

@jboynyc Your thoughts, please.

**Comment by sarahec** [comment]
FYI, the following updated build file generates a usable package. It:

1. Bumps to 0.9.4 (the latest in PyPi) to pick up your deprecation fixes
2. Uses the `relaxDependencyHook` to allow operation with current igraph and leidenalg
3. Ignores the `UserWarning` emitted by `network.py` during testing. Importing the Cythonized code works after the package is installed in `/nix/store/` but the .so file isn't available at test time. 
4. Skips patching the dependencies -- I'd forgotten about `relaxDependencyHook`

I've submitted this for consideration in nixpkgs to fix a broken build. Let me know if I should bump to 0.9.5 instead.

```nix
{ lib
, buildPythonPackage
, cairocffi
, cython_3
, fetchPypi
, igraph
, leidenalg
, pandas
, poetry-core
, pytestCheckHook
, pythonOlder
, pythonRelaxDepsHook
, scipy
, setuptools
, spacy
, spacy-lookups-data
, en_core_web_sm
, toolz
, tqdm
, wasabi
}:

buildPythonPackage rec {
  pname = "textnets";
  version = "0.9.4";
  format = "pyproject";

  disabled = pythonOlder "3.8";

  src = fetchPypi {
    inherit pname version;
    hash = "sha256-4154ytzo1QpwhKA1BkVMss9fNIkysnClW/yfSVlX33M=";
  };

  # TODO Remove dependency relaxation with version >= 0.9.5
  nativeBuildInputs = [
    pythonRelaxDepsHook
    cython_3
    poetry-core
    setuptools
  ];

  pythonRelaxDeps = [ "igraph" "leidenalg" ];

  propagatedBuildInputs = [
    cairocffi
    igraph
    leidenalg
    pandas
    scipy
    spacy
    spacy-lookups-data
    toolz
    tqdm
    wasabi
  ];

  nativeCheckInputs = [
    pytestCheckHook
    en_core_web_sm
  ];

  pythonImportsCheck = [
    "textnets"
  ];

   # The package has a few deprecation warnings that are thrown during testing (including as furure warnings).
   #
   # network.py:45 throws a UserWarning when it can't import Cythonized code. 
   # The .so files aren't present during the build, so let network.py use its pure-python replacement in testing.
   # These files will be present when the package is installed.
  pytestFlagsArray = [
    "-W" "ignore::DeprecationWarning" 
    "-W" "ignore::FutureWarning" 
    "-W" "ignore::UserWarning" 
  ];

  meta = with lib; {
    description = "Text analysis with networks";
    homepage = "https://textnets.readthedocs.io";
    changelog = "https://github.com/jboynyc/textnets/blob/v${version}/HISTORY.rst";
    license = licenses.gpl3Only;
    maintainers = with maintainers; [ jboy ];
  };
}
```

**Comment by sarahec** [comment]
Apologies for all the detail, I hadn't realized you were the maintainer! Happy to help in whatever way is best.

**Comment by jboynyc** [comment]
Thanks for all this @sarahec (and sorry for not reacting sooner). I hope to catch up on all this soon. I usually try to keep the version constraints somewhat aligned with what is in nixpkgs but clearly failed to do so on the latest release. I'll try to publish a release with more relaxed version constraints, that should ease the packaging.

**Comment by sarahec** [comment]
It's understandable; we're all busy. I have a PR under review for a 0.9.4 bump. It needed an interesting change for Cython.
